// react-static-plugin-typescript/node.api.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = require("fs");

var _path = _interopRequireDefault(require("path"));

var _forkTsCheckerWebpackPlugin = _interopRequireDefault(require("fork-ts-checker-webpack-plugin"));

var _convertTsconfigPathsToWebpackAliases = _interopRequireDefault(require("convert-tsconfig-paths-to-webpack-aliases"));

var _typescript = require("typescript");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var _default = function _default(options) {
  return {
    afterGetConfig: getReactStaticConfigExtender(),
    webpack: webpack(options)
  };
};

exports.default = _default;

function webpack(options) {
  var tsconfigLocation = (0, _typescript.findConfigFile)('./', _typescript.sys.fileExists);
  var tsconfigPath = tsconfigLocation ? _path.default.resolve(tsconfigLocation) : tsconfigLocation;
  var parsedTsConfig = tsconfigPath ? (0, _typescript.parseConfigFileTextToJson)(tsconfigPath, (0, _fs.readFileSync)(tsconfigPath, 'utf8')) : {
    config: {}
  };
  var tsAliases = parsedTsConfig.config.compilerOptions && parsedTsConfig.config.compilerOptions.paths ? (0, _convertTsconfigPathsToWebpackAliases.default)(parsedTsConfig.config) : {};
  return function (previousConfig, state) {
    var previousModuleConfig = previousConfig.module || {};
    var previousResolveConfig = previousConfig.resolve || {};
    var previousResolveExtensions = previousResolveConfig.extensions || [];
    var typescriptLoader = getTypeScriptLoader(state.defaultLoaders.jsLoader);
    var plugins = previousConfig.plugins || [];

    if ((options || {}).typeCheck !== false) {
      plugins.push(getTypecheckPlugin(tsconfigPath));
    }

    var loaders = Array.isArray(previousModuleConfig.rules) && previousModuleConfig.rules.length > 0 && Array.isArray(previousModuleConfig.rules[0].oneOf) ? _toConsumableArray(previousModuleConfig.rules[0].oneOf) : [state.defaultLoaders.jsLoader, state.defaultLoaders.jsLoaderExt, state.defaultLoaders.cssLoader, state.defaultLoaders.fileLoader];

    if (loaders.indexOf(state.defaultLoaders.jsLoader) !== -1) {
      // If the default Javascript loader is still present, replace it with the TypeScript loader:
      loaders[loaders.indexOf(state.defaultLoaders.jsLoader)] = typescriptLoader;
    } else {
      // Otherwise just add the TypeScript loader to the list of loaders, before the others:
      loaders.unshift(typescriptLoader);
    }

    return _objectSpread({}, previousConfig, {
      plugins: plugins,
      resolve: _objectSpread({}, previousResolveConfig, {
        extensions: previousResolveExtensions.concat(['.ts', '.tsx']),
        alias: _objectSpread({}, previousResolveConfig.alias, tsAliases)
      }),
      module: _objectSpread({}, previousModuleConfig, {
        rules: [{
          oneOf: loaders
        }]
      })
    });
  };
}

function getReactStaticConfigExtender() {
  return function (state) {
    return _objectSpread({}, state, {
      config: _objectSpread({}, state.config, {
        extensions: state.config.extensions.concat(['.ts', '.tsx'])
      })
    });
  };
}

function getTypeScriptLoader(jsLoader) {
  var jsLoaderBabelPresets = jsLoader.use[0].options.presets || [];

  var typescriptLoader = _objectSpread({}, jsLoader, {
    test: /\.(js|jsx|ts|tsx)$/,
    use: [_objectSpread({}, jsLoader.use[0], {
      options: _objectSpread({}, jsLoader.use[0].options, {
        presets: jsLoaderBabelPresets.concat('@babel/preset-typescript')
      })
    })]
  });

  return typescriptLoader;
}

function getTypecheckPlugin(tsconfigPath) {
  return new _forkTsCheckerWebpackPlugin.default({
    async: false,
    checkSyntacticErrors: true,
    tsconfig: tsconfigPath
  });
}
